Differentiable error mitigation

Qottmann opened the pull request:
A simple notebook showing off the new differentiable error mitigation functions with ZNE https://github.com/PennyLaneAI/pennylane/pull/2757 as well as trying to initiate people also to think of variationally optimizing the mitigation scheme itself.

Initial commit only contains a rough sketch of the contents to enable early feedback on what the demo should contain.

josh146 commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,60 @@
+
+Showing off new functionality with ZNE
+--------------------------------------
+Could be nice to base the tutorial on jax as the new functions are jitable and super fast?
```
Yes! Great idea! 💯

josh146 commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,60 @@
+Could be nice to base the tutorial on jax as the new functions are jitable and super fast?
+Including the example from the fold_global details section, showing the manual fit in ZNE to assure it makes sense (visualization always helps imo)
+
+VQE example using a noisy device with mitigation
+------------------------------------------------
```
Were you thinking of using an IBM noise model here? Or a different approach for inserting noise?

Qottmann replied:
The easiest path would be just mixed device with some noise channels like `qml.DepolarizingChannel` Obtaining the IBM noisy device would require loading IBMQ credentials and might make it more complicated than necessary. I think for ZNE our noise channels should suffice. What do you think? Do you have a preference?

josh146 replied:
Sounds good! Was just curious :)

josh146 commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,60 @@
+VQE example using a noisy device with mitigation
+------------------------------------------------
+
+Showing off other interfaces and how to get their gradients
+-----------------------------------------------------------
```
A random thought, but makes me wonder if this would be useful here: https://github.com/executablebooks/sphinx-tabs

Qottmann replied:
something like that would be very nice! This would require installing this package on the build platform right?

josh146 replied:
Sounds good! Was just curious :)

josh146 replied:
Yes, it would require a separate PR to setup. Out of scope for now, just a rabbit hole I almost went down!

Qottmann commented:
Is there a way to have fancy latex symbols like `\lightning`, in the demo? E.g. using `\usepackage{wasysym}` or `stmaryrd`.
I think it makes a lot of sense here and would make it much more readable (or maybe I just like the symbol 😄 )

josh146 commented:
@Qottmann you could use emojis (⚡) directly in the demo! Simply by copying and pasting the emoji directly into the text, or by using a VSCode plugin that allows you to insert emojis, should work.

josh146 commented:
Oh do you need to use the lightning symbol inside larger latex mathematical equations?

Qottmann commented:
Yes, and it is working!

Qottmann commented:
Do I have to manually tell the repo to use the latest master of pennylane? The deployment fails due to an `ImportError: cannot import name 'fold_global' from 'pennylane.transforms'`

josh146 commented:
@Qottmann you need to change the base of the PR to point to the `dev` branch:

- the `master` branch of the QML repo is pinned to the _released_ version of PL
- the `dev` branch is pinned to the GitHub master version of PL

![image](https://user-images.githubusercontent.com/2959003/178882130-43615469-1335-4b85-b19d-326863a47c11.png)

I just changed this for you, so retriggering the build should now work :)

josh146 commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -216,14 +216,15 @@ def VQE_run(cost_fn, max_iter, stepsize=0.4):
 grad_jax_jit = jax.jit(jax.grad(qnode_mitigated, argnums=[0, 1]))
-print(grad_jax_jit(w1_jax, w2_jax))
-print("Execution time: {} s".format(timeit.timeit("grad_jax_jit(w1_jax, w2_jax)", number = 100)/100))
+exec_time_jax_jit = timeit.timeit(lambda : grad_jax_jit(w1_jax, w2_jax), number = 100)/100
+print("Execution time jax jit grad: {} s".format(exec_time_jax_jit))
```
I'm not aware of the nuance, but in the JAX documentation they use `block_until_ready()` 🤔 

https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html#jit-compiling-a-function

KetpuntoG commented:
> Hello @KetpuntoG and @CatalinaAlbornoz , I tagged you for a review for this demo in which we want to showcase new functionalities added in [PennyLaneAI/pennylane#2757](https://github.com/PennyLaneAI/pennylane/pull/2757) as well as initiate discussion about differentiating mitigation parameters themselves. Do you think it is realistic to have this demo ready for publication with the new release, which includes these new features?

Yes, I think it is possible!
 It's going to take some work, but it makes sense to have it ready 🚀

josh146 commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,251 @@
+
+    tutorial_error_mitigation Error mitigation with Mitiq and PennyLane
+
+*Author: KK,.. Posted: 27 July 2022*
+
```
I suggest having 1-3 sentences here, before the first section starts, just as a brief introduction to let people know what the tutorial is going to do. E.g.,

> Error mitigation is an important strategy for minimizing noise when using noisy-intermediate scale (NISQ) quantum hardware. This can be very important when designing and testing variational algorithms. In this demo, we'll show how error mitigation can be combined with variational workflows, allowing you to differentiate _through_ the error mitigation.

[or something like that, feel free to re-word as needed!]

Qottmann commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,251 @@
+
+In order to run our VQA with our mitigated quantum function, we need to ensure that :math:`\tilde{f}` is differentiable - both formally and practically in our implementation.
+PennyLane now provides one such differentiable quantum error mitigation technique with `zero noise extrapolation` (ZNE), which can be used and differentiated in simulation and on hardware.
+We will briefly introduce these new functionalities and afterwards go more in depth to explore what happens under the hood.
```
I added a short sentence "Thus, we can improve the estimates of observables without breaking the differentiable workflow of our variational algorithm." before "We will briefly..."

josh146 replied:
Nice!

Qottmann commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,253 @@
+#
+# So far we have been using PennyLane gradient methods that use ``autograd`` for simulation and ``parameter-shift`` rules for real device
+# executions. We can also use the other interfaces that are supported by PennyLane, ``jax``, ``torch`` and ``tensorflow``, in the usual way 
+# as described in the :doc:`interfaces section of the documentation </introduction/interfaces>`.
```
Link still doesnt work. Tried every combination I could think of or find in other demos 🤔

josh146 replied:
Weird 🤔 I looked in a couple of other demos and saw

```
:doc:`introduction/optimizers`
```

and 

```
:doc:`parameter-shift rule </glossary/parameter_shift>`
```

Qottmann replied:
`:doc:`introduction/interfaces` works now 👍

josh146 replied:
Nice! I always get confused what the correct syntax is for Sphinx links

Qottmann commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,251 @@
+
+.. meta::
+    :property="og:description": Differentiable error mitigation
+    :property="og:image": https://pennylane.ai/qml/_images/diffable_mitigation_thumb.png
```
I uploaded the file to `_build/_images/diffable_mitigation_thumb.png` - is that not correct? Or is that a caching problem of the build? (It is still using the one which I used as a template to start from)

josh146 replied:
@Qottmann this is a subtlety of using Sphinx :)

Sphinx automatically copies all images you have in your `demonstrations/demo_name/` directory to `_images/` when building the website. So no need to place anything manually in `_images`. Instead, just use `https://pennylane.ai/qml/_images/image_filename.png` where `image_filename.png` exists _somewhere_.

Once the demo is merged, this link will start working.

Qottmann replied:
So this means the correct version can only be visible _after_ merging?
I changed it now and uploaded the thumbnail to the demo folder. Still weird to me why it would display the noisy cirq demo's thumbnail 🤔

josh146 replied:
Oh this isn't thumbnail, this is the _image metadata_. Basically, what appears when you post the link on social media (e.g., twitter). There are nice tools for checking this: https://cards-dev.twitter.com/validator (but this won't work until merging).

Let me show you where the thumbnail is set.

josh146 commented on demos_getting-started.rst:
```diff
@@ -94,6 +94,12 @@ Here you can discover the basic tools needed to use PennyLane through simple dem
     :description: :doc:`demos/tutorial_noisy_circuit_optimization`
     :tags: cirq
 
+.. customgalleryitem::
+    :tooltip: Learn how to use zne error mitigation and maintain differentiability.
+    :figure: demonstrations/noisy_circuit_optimization/noisy_circuit_optimization_thumbnail.png
+    :description: :doc:`demos/tutorial_diffable-mitigation`
+    :tags: mitigation zero noise extrapolation differentiability
```
@Qottmann _this_ (`:figure:`) is where the thumbnail is defined 🙂

Qottmann replied:
Oooops 🦆

KetpuntoG commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+
+Most variational quantum algorithms (VQAs) are concerned with optimizing a `quantum function` 
+
+.. math:: f(\theta) = \langle 0 | U^\dagger(\theta) H U(\theta) | 0 \rangle
```
Since just below you define the function from the trace it would be interesting to put the equivalent form of the trace in f to see exactly the difference between f and f⚡️

Qottmann replied:
Thanks for the point, my understanding is that you want to be sure no one is lost that isnt familiar with mixed states and their expectation values $\text{tr}(\rho H)$ versus the standard pure state formulation $\langle \psi | H | \psi \rangle = \text{tr}(|\psi \rangle \langle \psi | H)$? I dont think defining $f$ in terms of the trace formula is very helpful, so I tried to make it a bit clearer in the definition of f⚡️.

KetpuntoG replied:
yes, that's the idea 💪

KetpuntoG commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+Thus, we can improve the estimates of observables without breaking the differentiable workflow of our variational algorithm.
+We will briefly introduce these new functionalities and afterwards go more in depth to explore what happens under the hood.
+
+We start by initializing a noisy device under the ``qml.DepolarizingChannel``:
```
Add link to doc

KetpuntoG commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+
+n_layers = 2
+
+w1 = np.ones((n_wires), requires_grad=True)
```
I would clarify that these are the parameters of the model and you are giving them a random value, or that you are giving them the value 1

KetpuntoG commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+##############################################################################
+# Under the hood of Zero Noise Extrapolation
+# ------------------------------------------
+# What is happening here under the hood? The basic idea of ZNE is to artificially increase the noise in a circuit,
```
Oh that's cool! 😄

KetpuntoG commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+# We can then simply transform the noisy qnode :math:`f^{⚡}` with :func:`~.pennylane.mitigate_with_zne` to generate :math:`\tilde{f}`.
+# If everything goes as planned, executing the mitigated ``qnode`` is then closer to the ideal result:
+
+scale_factors = [1.0, 2.0, 3.0]
```
I understand that they can only be integers. It would be better to remove the dot

KetpuntoG commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+# The cool thing about this new mitigated QNode is that it is still differentiable! That is, we can compute its gradient as usual:
+
+grad = qml.grad(qnode_mitigated)
+print(grad(w1, w2))
```
this print is left with lateral displacement and in a not very pretty way. Perhaps you could make `print(grad(w1, w2)[0])` that with that already it is seen that it works and aesthetically it would be better

Qottmann replied:
Again not sure what is meant by lateral displacement? 🤔 
![image](https://user-images.githubusercontent.com/43949391/183700239-8e61d5d5-77f8-4ebd-afb0-291a72a17306.png)

KetpuntoG replied:
<img width="1193" alt="Captura de pantalla 2022-08-10 a las 11 21 04" src="https://user-images.githubusercontent.com/65235481/183942941-7bf3de51-2101-4b3e-a636-5cd542495e95.png">

KetpuntoG replied:
I am at 100% size, it seems it depend on the browser. It may be better to reduce just in case

Qottmann replied:
at this point of the code I also get the scrolling, I will change that! Do you still get it at this point here though?

KetpuntoG replied:
Yes, here too 😅

Qottmann replied:
What browser and device are you using? Are you on a tablet? I tried with Firefox and Chrome and varying zoom

KetpuntoG commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+# Note that this folding scheme is relatively simple and only really is sensible for integer values of ``scale_factor``. At the same time, ``scale_factor`` is
+# limited from above by the noise as the noisy quantum function quickly decoheres under this folding. Therefore, one typically only uses ``scale_factors = [1, 2, 3]``, but you
+# can in principle think of more fine grained folding schemes and test them by providing custom folding operations, see details in :func:`~.pennylane.transforms.mitigate_with_zne`.
+#
```
I have missed a small comment to know what is richardson_extrapolate or comment that there are more types

Qottmann replied:
Good point!

KetpuntoG commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+plt.xlabel("epoch", fontsize=18)
+plt.ylabel("energy", fontsize=18)
+plt.show()
+
```
Amazing plot! 😯

KetpuntoG reviewed (commented):
Really interesting! Very good job @Qottmann 😊
Here I left some comment

Qottmann commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+##############################################################################
+# We can then simply transform the noisy qnode :math:`f^{⚡}` with :func:`~.pennylane.mitigate_with_zne` to generate :math:`\tilde{f}`.
+# If everything goes as planned, executing the mitigated ``qnode`` is then closer to the ideal result:
+
```
What do you mean by lateral displacement? In my browser is looks normal 🤔 
![image](https://user-images.githubusercontent.com/43949391/183699917-cc63e6b2-3f65-4d72-afb5-5eecd022a212.png)

Qottmann replied:
can you make a screenshot here? Then I can roughly guess where to make changes. Its a bit strange though since this is just continuous text

KetpuntoG replied:
<img width="782" alt="Captura de pantalla 2022-08-10 a las 17 08 35" src="https://user-images.githubusercontent.com/65235481/184020440-4c08ac0c-abf9-4c44-b581-3c47e783f207.png">

KetpuntoG replied:
the cell below

KetpuntoG commented:
I just found this [demo](https://pennylane.ai/qml/demos/tutorial_error_mitigation.html), it would be good to quote it at some point

Qottmann commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+
+import pennylane as qml
+import pennylane.numpy as np
+from pennylane.transforms import mitigate_with_zne, fold_global, richardson_extrapolate
```
I just remove the import and call `qml.transforms.fold_global` instead now. I am also not a fan of having a lot of imports in the beginning 👍

Qottmann commented:
This demo here (kind of) outdates this older demo, as it showcases the same functionality, but now with proprietary PennyLane functions that can be differentiated. This other demo is also more a `mitiq` than a PennyLane demo. I dont think it should be removed, but I would not advertise it here because it could be confusing. What do you think?

KetpuntoG commented:
> This demo here (kind of) outdates this older demo, as it showcases the same functionality, but now with proprietary PennyLane functions that can be differentiated. This other demo is also more a `mitiq` than a PennyLane demo. I dont think it should be removed, but I would not advertise it here because it could be confusing. What do you think?

Ah great, glad to hear you had your eye on it. Forget it then 😄

Qottmann commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+#     `arXiv:2112.05821 <https://arxiv.org/abs/2112.05821>`__, 2021.
+
+##############################################################################
+#.. bio:: Korbinian Kottmann
```
Indeed, any idea how to fix that?

josh146 replied:
@Qottmann you need to put the image in the top level `_static/authors` folder :) Then you can do
```rest
:photo: ../_static/authors/qottman.jpg
```

Qottmann commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+
+##############################################################################
+# We can then simply transform the noisy qnode :math:`f^{⚡}` with :func:`~.pennylane.mitigate_with_zne` to generate :math:`\tilde{f}`.
+# If everything goes as planned, executing the mitigated ``qnode`` is then closer to the ideal result:
```
Good point!

KetpuntoG commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,257 @@
+
+##############################################################################
+#.. bio:: Korbinian Kottmann
+#    :photo: demonstrations/diffable-mitigation/qottmann.jpg
```
```suggestion
#    :photo: /demonstrations/diffable-mitigation/qottmann.jpg
```

I think this is the problem

Qottmann commented:
The last two deployments for some reason did not render my changes, is this a new CI feature? I.e. I changed some formatting but not the code itself, is it somehow recognizing that and hence not re-running the demo?

KetpuntoG commented:
From my side this is perfect, thank you very much for the changes!
We will carry out a last review and if all goes well next week it will be published 😄

isaacdevlugt commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,271 @@
+# In this mitigation technique, the single qubit state is put into equal superposition
+# :math:`|+\rangle = (|0\rangle + |1\rangle)/\sqrt{2}`. During the first idle time :math:`t_1`, the state is altered due to noise.
+# Applying :math:`X` reverses the roles of each computational basis state. The idea is that the noise in the second idle time
+# :math:`T-t_1` is canceling out the effect of the first time window. We see that the output fidelity with respect to the noise-free
```
> We see that the output fidelity with respect to the noise-free execution is a smooth function of ...

Where am I looking?

isaacdevlugt commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,271 @@
+# execution is a smooth function of :math:`t_1`. This was executed on ``ibm_perth``, and we note that simple noise models,
+# like the simulated IBM device, do not suffice to reproduce the behavior of the real device.
+#
+# Obtaining the gradient with respect to this parameter is difficult. There are formal and practical obstacles:
```
```suggestion
# Obtaining the gradient with respect to this parameter is difficult. 
```

isaacdevlugt commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,271 @@
+# Obtaining the gradient with respect to this parameter is difficult. There are formal and practical obstacles:
+# Formally, writing down the derivative of this transform with respect to the idle time in order to derive its parameter-shift
+# rules would require access to the noise model.
+# This, on the other hand, is very difficult for a realistic scenario. Further, most mitigation parameters are integers and would have
```
```suggestion
# This is very difficult for a realistic scenario. Further, most mitigation parameters are integers and would have
```

isaacdevlugt commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,271 @@
+#
+# Overall, this is a nice example of a mitigation scheme where varying the mitigation parameter has direct impact to the simulation result.
+# It is therefore desirable to be able to optimize this parameter at the same time as we perform a variational quantum algorithm.
+#
+#
+# References
```
Needs a conclusions section!

larawatson replied:
I noticed in the last two sections "so far" and "with respect to" are used multiple times, perhaps there is another way to say those things to reduce the repetition

Qottmann replied:
I changed one "so far" and can scratch one "with respect to" but I am having a hard time expressing "derivative with respect to x" other than that? Any suggestions?

Qottmann replied:
(gradient = derivatives)

larawatson replied:
@Qottmann "concerning" would be the other word I would think of; however, if "with respect to" makes more sense scientifically, it's good to leave it as-is.

isaacdevlugt reviewed (dismissed):
Well done @Qottmann! Very minor comments overall. Some parts need some clarification or summaries, and a few should be more concise. 

Would love to take another look when you're done going through it again!

P.S.

One thing that kind of "urks" me throughout the demo is when `scale_factor` is referred to in the text and code, but it's \lambda in the math (and sometimes you mix the two together). What are your thoughts on removing "scale_factor" everywhere and just calling and referring to it as "lambda"?

Qottmann commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,271 @@
+
+.. math:: f(\theta) = \langle 0 | U^\dagger(\theta) H U(\theta) | 0 \rangle
+
+for some Ansatz unitary :math:`U` with variational parameters :math:`\theta` and observable :math:`H`. These algorithms are specifically designed
+to be executed on noisy hardware. This means that naturally we do not have direct access to :math:`f`, but rather a noisy version :math:`f^{⚡}` where the variational state
```
> I don't think that's necessarily true? Maybe it can be re-worded as follows
I think its true but maybe not necessary to make such a strong statement 👍 

> Is "ansatz" a proper name? 🤔 I.e., should it be capitalized?
I capitalize it because nouns in German are capitalized in general. But now that I think about it, it probably does not translate to capitalizing German nouns in English 😄

josh146 replied:
> I don't think that's necessarily true? Maybe it can be re-worded as follows

Perhaps could be re-worded along the lines of "These algorithms arose due to the constraints of noisy near-term quantum hardware."?

> But now that I think about it, it probably does not translate to capitalizing German nouns in English 😄

I think I've seen both capitalized and non-capitalized, although the non-capitalized version I would say is much more common in English (see the [Wikipedia page](https://en.wikipedia.org/wiki/Ansatz)).

Qottmann commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,271 @@
+Thus, we can improve the estimates of observables without breaking the differentiable workflow of our variational algorithm.
+We will briefly introduce these new functionalities and afterwards go more in depth to explore what happens under the hood.
+
+We start by initializing a noisy device under the :func:`~.pennylane.DepolarizingChannel`:
```
Honestly, just to have "some" noise in there, the depolarizing one seems quite standard. What kind of argument would you like to see here? Being "common" to use seems not very convincing / worth adding.

Qottmann replied:
Anything I came up with to add here creates more confusion than clarifies anything so I would just leave it as is

Qottmann commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,271 @@
+# We see that the mitigated result comes close to the ideal result, whereas the noisy result is further off (see value at ``scale_factor=1``).
+#
+# Note that this folding scheme is relatively simple and only really is sensible for integer values of ``scale_factor``. At the same time, ``scale_factor`` is
+# limited from above by the noise as the noisy quantum function quickly decoheres under this folding. Therefore, one typically only uses ``scale_factors = [1, 2, 3]``, but you
```
Increasing `scale_factor` will create more noisy and noisy executions that will lead to complete decoherence, in many cases already for `scale_factor >= 4`. I can try to rewrite this!

Qottmann commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,271 @@
+##############################################################################
+# We see that the mitigated result comes close to the ideal result, whereas the noisy result is further off (see value at ``scale_factor=1``).
+#
+# Note that this folding scheme is relatively simple and only really is sensible for integer values of ``scale_factor``. At the same time, ``scale_factor`` is
```
No not really, what I mean would be coming up with a whole new folding scheme. The `fold_global` is just one of many ways to do this. It is not particularly fine but empirically it has been shown to work well, i.e. that lambda actually does scale with the noise (which is not super obvious for other folding schemes).

Qottmann replied:
I reformulated this part

isaacdevlugt commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,271 @@
+# and then extrapolate to :math:`\lambda \rightarrow 0`. Note that ``scale_factor=1`` corresponds to the original circuit, i.e. the noisy execution.
+#
+
+scale_factors = [1, 2, 3]
```
>Our measurement should depend polynomially on :math:\lambda

This itself is a little nuanced... Anyway, you don't have to use anything I wrote ^. But 1-2 sentences would be great!

Qottmann replied:
Similar situation here, I feel like these things are self-explanatory and adding a description in explicit words creates more confusion than it helps. I added a brief sentence that the extrapolation is done with a polynomial fit in f* vs $\lambda$ 👍

Qottmann commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,271 @@
+# limited from above by the noise as the noisy quantum function quickly decoheres under this folding. Therefore, one typically only uses ``scale_factors = [1, 2, 3]``, but you
+# can in principle think of more fine grained folding schemes and test them by providing custom folding operations, see details in :func:`~.pennylane.transforms.mitigate_with_zne`.
+#
+# Note that Richardson extrapolation, which we used to define the ``mitigated_qnode``, is just a fancy way to describe a polynomial fit of ``order = len(x) - 1``. Alternatively,
```
the x axis?

larawatson commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,269 @@
+
+*Author: Korbinian Kottmann, Posted: 27 July 2022*
+
+Error mitigation is an important strategy for minimizing noise when using noisy-intermediate scale quantum (NISQ) hardware.
```
Error mitigation is an important strategy for minimizing noise when using noisy-intermediate scale quantum (NISQ) hardware, especially when designing and testing variational algorithms.

josh146 replied:
```suggestion
Error mitigation is an important strategy for minimizing noise when using noisy-intermediate scale quantum (NISQ) hardware, especially when designing and testing variational algorithms.
```

larawatson commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,269 @@
+*Author: Korbinian Kottmann, Posted: 27 July 2022*
+
+Error mitigation is an important strategy for minimizing noise when using noisy-intermediate scale quantum (NISQ) hardware.
+This can be very important when designing and testing variational algorithms. In this demo, we will show how error mitigation
```
If we don't make the edit suggested above, I would recommend using something other than "important" here (just due to repetition)

larawatson commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,269 @@
+
+    tutorial_error_mitigation Error mitigation with Mitiq and PennyLane
+
+*Author: Korbinian Kottmann, Posted: 27 July 2022*
```
Perhaps not important and may happen automatically, but it may make sense to update the date to the release day (which I believe is set for next Tuesday, August 23rd)

Qottmann replied:
Thanks for pointing that out, I already changed it to 23rd of August but didnt update it online yet :)

Qottmann commented:
I could not get the suggested latex expression inside the matplotlib part rendered so I reversed that change. I could render it locally, so should be good now with the latest commit.

@isaacdevlugt I addressed your comments.

@larawatson Same, the only thing left is the "with respect to", not sure how to resolve that. Since we are talking about derivatives it would be okay for me to have "with respect to" appear multiple times in the last paragraph.

I also put it into "optimization" section since that felt more adequate (was "getting started" before). Let me know if you want to change that.

larawatson commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,289 @@
+# Conclusion
+# ----------
+# 
+# We demonstrated how zero-noise extrapolation can be seemlessly incorporated in a differentiable workflow in PennyLane to achieve better results.
```
```suggestion
# We demonstrated how zero-noise extrapolation can be seamlessly incorporated in a differentiable workflow in PennyLane to achieve better results.
```

KetpuntoG commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,289 @@
+
+
+##############################################################################
+# The quantum function, the expectation value of $H$, can then be executed on the noisy or ideal device
```
```suggestion
# The quantum function, the expectation value of :math:`H`, can then be executed on the noisy or ideal device
```

KetpuntoG commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,289 @@
+# We can then simply transform the noisy QNode :math:`f^{⚡}` with :func:`~.pennylane.transforms.mitigate_with_zne` to generate :math:`\tilde{f}`.
+# If everything goes as planned, executing the mitigated QNode is then closer to the ideal result:
+
+scale_factors = [1.0, 2.0, 3.0]
```
```suggestion
scale_factors = [1, 2, 3]
```

KetpuntoG commented on demonstrations/tutorial_diffable-mitigation.py:
```diff
@@ -0,0 +1,289 @@
+
+    tutorial_error_mitigation Error mitigation with Mitiq and PennyLane
+
+*Author: Korbinian Kottmann, Posted: 23 August 2022*
```
```suggestion
*Author: Korbinian Kottmann, Posted: 22 August 2022*
```

KetpuntoG reviewed (approved):
Great job, with this it would be ready!

isaacdevlugt commented:
Nice work @Qottmann 🥳
